sources/imdsclient/src/lib.rs

```diff
@@ -221,91 +240,91 @@ impl ImdsClient {
             target.as_ref()
         );
         debug!("Requesting {}", &uri);
-        let mut attempt: u8 = 0;
-        let max_attempts: u8 = 3;
-        loop {
-            attempt += 1;
-            ensure!(attempt <= max_attempts, error::FailedFetchIMDS { attempt });
-            if attempt > 1 {
-                time::sleep(Duration::from_secs(1)).await;
-            }
-            let response = self
-                .client
-                .get(&uri)
-                .header("X-aws-ec2-metadata-token", &self.session_token)
-                .send()
-                .await
-                .context(error::Request {
-                    method: "GET",
-                    uri: &uri,
-                })?;
-            trace!("IMDS response: {:?}", &response);
-
-            match response.status() {
-                code @ StatusCode::OK => {
-                    info!("Received {}", target.as_ref());
-                    let response_body = response
-                        .bytes()
-                        .await
-                        .context(error::ResponseBody {
-                            method: "GET",
-                            uri: &uri,
-                            code,
-                        })?
-                        .to_vec();
-
-                    let response_str = printable_string(&response_body);
-                    trace!("Response: {:?}", response_str);
-
-                    return Ok(Some(response_body));
-                }
-
-                // IMDS returns 404 if no user data is given, or if IMDS is disabled
-                StatusCode::NOT_FOUND => return Ok(None),
-
-                // IMDS returns 401 if the session token is expired or invalid
-                StatusCode::UNAUTHORIZED => {
-                    info!("Session token is invalid or expired");
-                    self.refresh_token().await?;
-                    info!("Refreshed session token");
-                    continue;
-                }
+        timeout(
+            self.retry_timeout,
+            Retry::spawn(retry_strategy(), || async {
+                let session_token = {
```

Comment:
I think this async block should be a function defined elsewhere and called here (for readability).